### PR TITLE
Add Ganons Dungeon Shuffle Option

### PR DESCRIPTION
**Pull request by Jaybone25**
Added new option to dungeon shuffle for Ganon + Dungeon.   This allows for an alternate win condition.

### CODE REVIEW

File: SettingsList.py [hunk 1]
```diff
@@ -3018,21 +3018,33 @@ def __init__(self, name, gui_text, min, max, default, step=1,
...
+            'off':		 'Off',
+            'simple':	 'Dungeon',
+            'all':		 'Dungeon and Ganon',
+        },
         gui_tooltip    = '''\
-            Shuffle the pool of dungeon entrances, including Bottom 
-            of the Well, Ice Cavern, and Gerudo Training Ground.
-            However, Ganon's Castle and Thieves' Hideout are not shuffled.
-
+            Dungeon Shuffle will the pool of dungeon entrances, including
```
Comment (fenhl) [inline SettingsList.py:3031]: missing a verb here

**Comment by r0bd0g** [comment]
I don't understand... How do you win the game now?

**Comment by r0bd0g** [comment]
Logic:
- You need to check for swords to fight the enemies in Ganon's Tower.
- You don't actually have to check for is_child when the adult condition is just is_adult b/c if you're not adult you must be child. I dunno it doesn't matter, just a few extra brackets floating around can get confusing sometimes.
- In vanilla Spirit trial, the checks there you need to can_use around the Hookshot or else child with the Hookshot can be expected to get the chests without the trick on.
- And then the Spirit Trial nut pot is not even accessible as child at all.
- And speaking of the Spirit Trial trick, it works as child but I believe, not 100% sure, that you must jumpslash to get the rupee without Hookshot.
- I don't like how the first two regions in the MQ Ganon file are in the wrong order.
- I think you've got a bracketing error on the is_child section of Lobby->Main in MQ Ganon's. I don't think it actually causes a problem even though the operator precedence causes it to actually be wrong, b/c again you don't actually have to check for is_child here. This line is really complicated though; we talked about it before but I forget what we decided... But I think I understand where this is coming from. Kokiri Sword beats everything but you need something extra for the Armos. Explosives does the Iron Knuckle but it's too many explosives to rely only on them so we added in something else for the rest of the enemies? I guess it makes sense.
- I wonder how necessary it is to defeat the Stalfos before boomeranging the freestanding key. Certainly it would be annoying to have to do it while they're alive. I wonder if you can stun them how much time that gives you. You could stun with nuts but you must have rang and that could probably stun them too. The sort of thing I would have to go in-game to see how it plays, but I'm not going to. Have you tried it? If it's bad I'll take your word for it for now.
- Not seeing any check for adult on the MQ Shadow Trial chests.
- Not seeing any check for adult on the MQ Spirit Trial checks. (You can put these onto the checks themselves but maybe you'd rather shortcut it and put the adult check on room access instead, keep it simple.)

**Comment by cjohnson57** [comment]
> I don't understand... How do you win the game now?
> 
> Is it still just beat Ganon at the end of Ganon's Castle? So what is the rainbow bridge for? If it's just to access a single entrance, I think you need to do a lot of extra work here, as far as like, reworking path hints, changing when tokens are major items, and maybe other things of that nature. Those changes seem like they should be required but they interact poorly in some cases like, what if Ganon's goes into Ganon's anyway? Now all the things we'd change b/c Ganon's isn't necessary in Ganon's, would be changed for no reason. The randomizer ATM is just not at all designed for the wincon to not be across that bridge. Most of the work to accept a PR like this IMO is actually on that front.

While I'm a bit skeptical as well, I can see this being a fun setting with open bridge and GBK wincons instead of bridge ones

**Comment by r0bd0g** [comment]
It might play better if you made it so that whatever is across that bridge is the end of the game, and ganon's is just another dungeon somehow. But that also has a lot of questions about how it ought to function.

**Comment by ETR-BTF** [comment]
> It might play better if you made it so that whatever is across that bridge is the end of the game, and ganon's is just another dungeon somehow. But that also has a lot of questions about how it ought to function.

You mean make whatever boss is at the end of the dungeon across the bridge warp you to the credits?

**Comment by Jaybone25** [comment]
Yeah mainly its meant for GBK wincons or open bridge.   It was a bit interesting with mixed as well when I was messing with it.   Thanks for all the tips for the logic Rob, Felix is working on those and I'll get it updated soon.

Hmm never considered the hints, I'll have to take a look at those, I'm not sure it will hint them because it would know that Ganons tower is moved to not behind the bridge, but should be easy to test.   If you have any other things that might be broke let me know.   Worst case this could just be edited to only be allowed with open bridge until the groundwork for the other parts is completed.

**Comment by r0bd0g** [comment]
If whatever's across the rainbow bridge is the wincon, then I guess there's a thought about how that could work:

- The exit from Ganon's Castle to Ganon's Tower is never shuffled. Ganon's Tower to Ganondorf fight is added to the boss room entrance pool (not sure if this should only occur if Ganon's Tower is shuffled). (Some hint changes will be needed because I guess 'path to Ganondorf' becomes a thing. I can't think of why it would be a problem but might have to check there are no weird interactions with 'path to Evil's Bane' for light arrows on vanilla bridge.)
- Even on mixed pools, a dungeon is forced into the Ganon's Castle entrance (and a boss is forced into the boss door of that dungeon even if boss room mixed pools becomes a thing).
- When you defeat Ganondorf, after his death cutscene, you get loaded back into the Ganondorf boss room, the boss defeated fanfare plays, and a blue warp and heart container item spawn somewhere in the room. Stepping into the warp takes you to chamber of sages I guess where Zelda I guess gives you a stone/medallion I guess. If you return to the boss room, Ganondorf stays defeated. (Sort of strange side effect here is that Ganondorf will hint the light arrows even though they're not necessarily required to beat the game anymore. Not really sure what if anything should be done about that. I could see some weird cases where people go to Ganondorf dungeon to get a light arrows hint so they can clear trials.)
- When you beat the boss of that dungeon inside Ganon's entrance it sets the dungeon complete flags but the boss room ignores them so that the boss respawns every time you enter, and at the end of the boss death cutscene you are automatically warped to either Tower Collapse or the Ganon fight (depends on whether collapse is skipped).
- Savewarping out of Tower Collapse/Ganon takes you to the entrance of the dungeon inside Ganon's entrance.

This spec reads like a fanfiction; there's basically no chance it gets implemented this way lol.

**Comment by fenhl** [comment]
I think it would make more sense to keep defeating Ganondorf as the wincon. All the changes to make Ganondorf work like a regular boss and the boss in the castle work like Ganondorf sound like too much work to realistically happen any time soon. If and when someone does implement them, I still think they should be behind a setting.

I also want to point out that this setting would work particularly well with Triforce hunt. Ganon's castle pretty much just becomes a side dungeon in that setting, so treating it that way for ER purposes makes sense.

**Comment by r0bd0g** [comment]
I said before we've got a handful of things that would probably have to change if the rainbow bridge is not so important. The major itemness of tokens could be considered one of those things. But if Ganon's randomly goes back into Ganon's, it's awkward now that we've made those changes when it would maybe play better the way it was before.

**Comment by fenhl** [comment]
I don't think the Ganon's castle entrance being rolled vanilla should receive any special handling. We don't make tokens major items if something happens to be on a token reward either, for example.

**Comment by Jaybone25** [comment]
I updated logic per suggestions based off R0b.  I have also tested the interaction with Goal hints and it works as expected.   The path hints don't point at the bridge as the game knows that it isn't required in some seeds.   It does point at the bridge on the path of the tower if for example light arrows are in something past the bridge.

The MQ Forest trial free standing IS doable without anything, but it sucks.   Wouldn't want to require getting both of the stalfos stunned to get it

**Comment by Jaybone25** [comment]
I also played a seed and we used it as scrubs asyncs with it shuffled with 4 dungeon bridge and 6 med BK it plays pretty well with that

**Comment by r0bd0g** [comment]
Logic Review:
- I recommend some sort of tabbing on the MQ Ganon's Lobby->Main line.
- Do you know whether the eye switch in MQ Spirit Trial (that brings down the Iron Knuckle) is permanent or temporary? I would guess temporary but if you haven't checked, you should. (Hit the switch, then leave the dungeon and come back.) (I can't imagine destroying the thrones being permanent but I guess you could check that as well.)
- Worth noting for anybody else reading this they might not agree with putting is_adult on the trials access for MQ Spirit and Shadow. It works but it could be a trap for somebody later on down the line. That said it's pretty strange that you added is_adult onto the MQ Spirit Trial Nut Pot anyway, but only the nut pot (overcompensating for the lack of is_adult on the vanilla nut pot I think lol).
- Killing the MQ Forest Trial Stalfos is permanent, so you could kill them as adult and then boomerang the freestanding as child. The unfrozen eye switch in the following room, same thing, kill as adult, come back with slingshot. I called attention to this room before but I think I got distracted by wondering about whether you could boomerang the item that I forgot about the actual thing I wanted to say.
- In vanilla Spirit Trial, we have a logic helper that includes adult, k-sword, and sticks, I think it's can_jumpslash.
- I'm positive k sword has enough range to reach the crystal switch in vanilla Spirit Trial. Like I'm pretty sure I've seen that switch hit plenty of times with the k sword in speedruns or beta quests or whatever. Did you check?
- I guess you've accidentally backspaced some font colours and naming changes and things from the overworld file. Hopefully that doesn't get accidentally merged.
- You added the sword check in the tower for the vanilla BK chest. I'm not sure, I have no idea, if you need one for the Ganondorf hint or not. I think you actually do???? Imagine Ganon's Tower is child only and you have no K Sword. You need the light arrows to cross vanilla bridge to get an item to access the Tower as adult. The logic will think the light arrows hint is logically reachable even when it isn't, and therefore gossip stones will not be able to give you direct light arrows hints.

File: data/World/Overworld.json [hunk 2]
```diff
@@ -74,7 +74,6 @@
     },
     {
         "region_name": "Kokiri Forest",
-        "font_color": "Green",
```
Comment (fenhl) [inline data/World/Overworld.json:77]: Looks like this is reverting some of the changes from #1428. I'm guessing this is accidental?

File: World.py [hunk 3]
```diff
@@ -55,6 +55,9 @@ def __init__(self, id, settings, resolveRandomizedSettings=True):
 
         self.shuffle_special_interior_entrances = settings.shuffle_interior_entrances == 'all'
         self.shuffle_interior_entrances = settings.shuffle_interior_entrances in ['simple', 'all']
+		
+        self.shuffle_special_dungeon_entrances = settings.shuffle_dungeon_entrances == 'all'
+        self.shuffle_dungeon_entrances = settings.shuffle_dungeon_entrances in ['simple', 'all']
```
Comment (fenhl) [inline World.py:60]: The usage of `World.settings.shuffle_dungeon_entrances` in the remaining codebase should be replaced with `World.shuffle_dungeon_entrances` accordingly.

**Comment by mracsys** [comment]
Just chiming in to say I'd like to see this. If not in main, at least in one of the side branches.

From DMs, decoupled logic isn't handled yet. Adding a region across the rainbow bridge with just the dungeon entrance and an exit back to Castle Grounds also requiring `can_build_rainbow_bridge` should work.

**Comment by r0bd0g** [comment]
(BQ changed the spawn point for leaving Ganon's. Not sure if it only applies when the rainbow bridge isn't there or not.)

**Comment by fenhl** [comment]
Copying the modification from BQ would make sense. Does anyone know what exactly the modification is? I've been looking at the BQ code and I'm guessing it's [somewhere in here](https://github.com/mzxrules/Beta-Quest/blob/2709f7a018aa5bdd47117b373b7a2da86bf9f4e2/src/code.asm#L60-L77) but I don't know how to interpret that code.

**Comment by cjohnson57** [comment]
> Copying the modification from BQ would make sense. Does anyone know what exactly the modification is? I've been looking at the BQ code and I'm guessing it's [somewhere in here](https://github.com/mzxrules/Beta-Quest/blob/2709f7a018aa5bdd47117b373b7a2da86bf9f4e2/src/code.asm#L60-L77) but I don't know how to interpret that code.

https://www.eg.bucknell.edu/~csci320/mips_web/ This seems to work for converting the raw binary instruction into human readable instructions

**Comment by fenhl** [comment]
Are these even code? I'd expect entrance table entries to be data.

**Comment by cjohnson57** [comment]
oh, my bad. I saw "Modifications" and thought it was asm changes in raw binary format, you're probably right

**Comment by fenhl** [comment]
@cjohnson57 Would a change to handle the missing rainbow bridge, whether it's an adjustment to the entrance position or a logic fix, be acceptable in main Dev, or should it be maintained in Dev-R?

**Comment by r0bd0g** [comment]
I updated my last logic review to make it more clear where the actual bugs are.

**Comment by fenhl** [comment]
Another thing that should probably be added here is making the entrance behind the rainbow bridge an always hint with some bridge requirements.

**Comment by fenhl** [comment]
I'm currently waiting for @Jaybone25 to give me access to this branch so I can push a rebase onto latest Dev. Then I also have a branch ready to PR that adds an entrance hint for `Ganons Castle Grounds -> Ganons Castle Lobby` with the hint category depending on the bridge condition:

* `entrance_always` with vanilla, 2 or 3 stones, 2 or more medallions, 3 or more dungeon rewards, 21 or more tokens, or at least 2 more hearts than starting hearts
* `entrance` with 1 stone, 1 medallion, 2 dungeon rewards, 11 to 20 tokens, or 1 more heart than starting hearts
* Unhinted with open, 1 dungeon reward, 1 to 10 tokens, or a number of hearts covered by starting hearts

`entrance_always` is a new hint category that behaves like `always` or `dual_always`.

How do these conditions sound?

**Comment by r0bd0g** [comment]
I'm not sure I want to introduce the concept of an entrance always?
A lot of the conditions sound a lot easier to satisfy than finding three strength upgrades... I don't want OGC always hinted. So I worry about the slope and if it's slippery or not.

**Comment by fenhl** [comment]
> A lot of the conditions sound a lot easier to satisfy than finding three strength upgrades

I tried to match other sometimes/always hint conditions. 20 skulls is a sometimes hint and 30 is an always hint, for example, while 10 skulls isn't hinted, so for the bridge hint condition I went with 11–20 for sometimes and 21+ for always. Maybe these should be 15–24 and 25+ respectively. Checks that require dungeon rewards (OoT and Burning Kak) are always hints if they require at least 2 more rewards than the wincons, so I made the bridge entrance an always hint if at least 2 rewards are required.

**Comment by fenhl** [comment]
@cjohnson57 I think waiting for author can be removed now.

**Review by cjohnson57 — APPROVED** [review]

**Comment by cjohnson57** [comment]
Note for next time, when you change the values for a setting all unit tests need to be updated as well, I just did it myself this time though as I should have pointed it out earlier.

**Pull request merged**